dotnet/src/SemanticKernel/Orchestration/Plan.cs

```diff
@@ -0,0 +1,281 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.Text.Json.Serialization;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.SemanticKernel.AI.TextCompletion;
+using Microsoft.SemanticKernel.SkillDefinition;
+
+namespace Microsoft.SemanticKernel.Orchestration;
+
+/// <summary>
+/// Standard Semantic Kernel callable plan.
+/// Plan is used to create trees of <see cref="ISKFunction"/>s.
+/// </summary>
+public sealed class Plan : ISKFunction
+{
+    /// <summary>
+    /// State of the plan
+    /// </summary>
+    [JsonPropertyName("state")]
+    public ContextVariables State { get; } = new();
+
+    /// <summary>
+    /// Steps of the plan
+    /// </summary>
+    [JsonPropertyName("steps")]
+    internal List<Plan> Steps { get; } = new();
+
+    /// <summary>
+    /// Named parameters for the function
+    /// </summary>
+    [JsonPropertyName("named_parameters")]
+    public ContextVariables NamedParameters { get; set; } = new();
+
+    public bool HasNextStep => this.NextStep < this.Steps.Count;
+
+    public int NextStep { get; private set; } = 0;
```

Comment:
Change the name for better legibility and make it nullable when no steps are present.
```suggestion
    public int NextStepIndex { get; private set; } = 0;
```